source/common/stats/thread_local_store.cc

```diff
@@ -204,29 +255,130 @@ Histogram& ThreadLocalStoreImpl::ScopeImpl::histogram(const std::string& name) {
   // See comments in counter(). There is no super clean way (via templates or otherwise) to
   // share this code so I'm leaving it largely duplicated for now.
   std::string final_name = prefix_ + name;
-  HistogramSharedPtr* tls_ref = nullptr;
+  std::unique_lock<std::mutex> lock(parent_.lock_);
+  ParentHistogramImplSharedPtr& central_ref = central_cache_.histograms_[final_name];
+
+  std::vector<Tag> tags;
+  std::string tag_extracted_name = parent_.getTagsForName(final_name, tags);
+  if (!central_ref) {
+    // Since MetricImpl only has move constructor, we are explicitly copying here.
+    std::string central_tag_extracted_name(tag_extracted_name);
+    std::vector<Tag> central_tags(tags);
+    central_ref.reset(new ParentHistogramImpl(final_name, parent_, *this,
+                                              std::move(central_tag_extracted_name),
+                                              std::move(central_tags)));
+  }
+  return *central_ref;
+}
+
+Histogram& ThreadLocalStoreImpl::ScopeImpl::tlsHistogram(const std::string& name) {
+  // See comments in counter() which explains the logic here.
+
+  // Here prefix will not be considered because, by the time ParentHistogram calls this method
+  // during recordValue, the prefix is already attached to the name.
+  TlsHistogramSharedPtr* tls_ref = nullptr;
   if (!parent_.shutting_down_ && parent_.tls_) {
-    tls_ref = &parent_.tls_->getTyped<TlsCache>().scope_cache_[this].histograms_[final_name];
+    tls_ref = &parent_.tls_->getTyped<TlsCache>().scope_cache_[this].histograms_[name];
   }
 
   if (tls_ref && *tls_ref) {
     return **tls_ref;
   }
 
   std::unique_lock<std::mutex> lock(parent_.lock_);
-  HistogramSharedPtr& central_ref = central_cache_.histograms_[final_name];
-  if (!central_ref) {
-    std::vector<Tag> tags;
-    std::string tag_extracted_name = parent_.getTagsForName(final_name, tags);
-    central_ref.reset(
-        new HistogramImpl(final_name, parent_, std::move(tag_extracted_name), std::move(tags)));
-  }
+  std::vector<Tag> tags;
+  std::string tag_extracted_name = parent_.getTagsForName(name, tags);
+  TlsHistogramSharedPtr hist_tls_ptr = std::make_shared<ThreadLocalHistogramImpl>(
+      name, std::move(tag_extracted_name), std::move(tags));
+
+  ParentHistogramImplSharedPtr& central_ref = central_cache_.histograms_[name];
```

Comment:
Since this is always called from the parent histogram, could we just pass the parent in rather than looking it up in the cache?